variables: allow resolving `extensionDir`

connor4312 opened the pull request:
This allows us to fix https://github.com/microsoft/vscode-remote-release/issues/5516#issuecomment-911597917

It enables a new replacement in the format `${extensionDir:<id>}` which
will expand to the filesystem path where the extension is stored. This
involved churn, since now resolution is always synchronous (where before
the terminal took a synchronous-only path.)

Additionally, changes were needed to inject this information in the
variable resolver. As part of this I made the extension host resolver
(used by debug and tasks) its own extension host service.

connor4312 commented on src/vs/server/node/remoteTerminalChannel.ts:
```diff
@@ -222,7 +230,7 @@ export class RemoteTerminalChannel extends Disposable implements IServerChannel<
 			const envVariableCollections = new Map<string, IEnvironmentVariableCollection>(entries);
 			const mergedCollection = new MergedEnvironmentVariableCollection(envVariableCollections);
-			mergedCollection.applyToProcessEnvironment(env);
+			await mergedCollection.applyToProcessEnvironment(env, variableResolver);
```
We actually were not resolving variables in environment variable collections in remotes; passing the variableResolver fixes that

alexr00 reviewed (commented):
I tried reading through https://github.com/microsoft/vscode-remote-release/issues/5516, but there's a lot going on there. Where would a user use `${extensionDir:<id>}`?

connor4312 commented:
> I tried reading through [microsoft/vscode-remote-release#5516](https://github.com/microsoft/vscode-remote-release/issues/5516), but there's a lot going on there. Where would a user use `${extensionDir:<id>}`?

In js-debug, we add an environment variable to import the "bootloader" script automatically in Node.js programs to enable auto attach. This is only updated when the user changes their desired auto attach behavior; js-debug doesn't activate on `*` or eagerly check it. The bootloader is included in js-debug, and if the variable points to a location that doesn't exist, all Node.js programs will error out.

The issue the user ran into is that they updated to a new version of VS Code, and therefore the path to the old bootloader on their remote--which is set as an ordinary absolute path--was no longer valid. This can be fixed with the introduction of a `${extensionDir:}` variable that VS Code will evaluate when creating a new terminal, so that the location will always be right.

Tyriar commented on src/vs/workbench/services/configurationResolver/common/variableResolver.ts:
```diff
@@ -26,6 +27,7 @@ export interface IVariableResolveContext {
 	getWorkspaceFolderPathForFile?(): string | undefined;
 	getSelectedText(): string | undefined;
 	getLineNumber(): string | undefined;
+	getExtension(id: string): Promise<{ readonly extensionLocation: uri } | undefined>;
```
Should this be `getExtensionDir` instead?

connor4312 replied:
I named it like this in case we pull more stuff from the extension info at a later date. The return value from this is a subset of the data from `IExtensionService.getExtension`.

Tyriar commented on src/vs/workbench/contrib/terminal/browser/terminalProfileResolverService.ts:
```diff
@@ -355,25 +355,25 @@ export abstract class BaseTerminalProfileResolverService implements ITerminalPro
+				profile.args = await this._resolveVariables(profile.args, env, lastActiveWorkspace);
 			} else {
 				for (let i = 0; i < profile.args.length; i++) {
-					profile.args[i] = this._resolveVariables(profile.args[i], env, lastActiveWorkspace);
+					profile.args[i] = await this._resolveVariables(profile.args[i], env, lastActiveWorkspace);
```
Can we await Promise.all here?

Tyriar commented on src/vs/workbench/contrib/terminal/common/terminalEnvironment.ts:
```diff
@@ -78,17 +78,18 @@ function mergeNonNullKeys(env: IProcessEnvironment, other: ITerminalEnvironment
 		if (typeof value === 'string') {
 			try {
-				env[key] = variableResolver(value);
+				env[key] = await variableResolver(value);
```
await Promise.all? Worried about extending the launch time of shells

Tyriar reviewed (approved):
Terminal stuff looks good.

+ @sandy081 for the extension service

alexr00 commented:
> This can be fixed with the introduction of a ${extensionDir:} variable that VS Code will evaluate when creating a new terminal, so that the location will always be right.

Does this mean that a real user will never use `${extensionDir:}` in a launch.json file?

alexr00 commented on src/vs/workbench/services/configurationResolver/common/variableResolver.ts:
```diff
@@ -268,6 +271,16 @@ export class AbstractVariableResolverService implements IConfigurationResolverSe
+					if (!ext) {
+						throw new Error(localize('extensionNotInstalled', "Variable {0} can not be resolved because the extension {1} is not installed.", match, argument));
+					}
+					return this.fsPath(ext.extensionLocation);
```
This is fine, but we should make sure it's well documented in the variable reference doc page.

connor4312 commented:
> Does this mean that a real user will never use ${extensionDir:} in a launch.json file? I ask because these variables should all be things that real users want to use in tasks, debug, and maybe settings.

Hm, they _could_ be if the user ever needed to reference a file in an extension. Though that is rare and I don't know of other cases yet. More likely would be other extensions using this (or generating this) like js-debug does.

alexr00 commented:
@connor4312 having a variable that we don't expect any users to actually use is not ideal. You did a lot of great work in this PR, but it's not clear to me that having a variable is the best way to solve https://github.com/microsoft/vscode/pull/146274#issuecomment-1083302596 (thank you for the explanation btw!). 

Since this is something that only extension would use, it is possible to solve it with API?

connor4312 commented:
Hmm, maybe. An API-specific substitution or something... @Tyriar any thoughts?

Tyriar commented:
How about prefixing with `__` and just not documenting it? It would be a shame to have a whole new variable resolver mechanism that we then need to initialize on all these different processes it works on.

sandy081 commented:
Transferring to @alexdima as he owns extension service

alexr00 commented:
> How about prefixing with __ and just not documenting it?

If we do this then it becomes defacto API. Could we make an API that just gets the install location of an extension without doing any substitution and then pass that to whatever terminal API needs it?

alexdima reviewed (approved):
:+1:

alexr00 commented:
Reaching back to an older comment:

> The issue the user ran into is that they updated to a new version of VS Code, and therefore the path to the old bootloader on their remote--which is set as an ordinary absolute path--was no longer valid.

Was this they updated their extension, not VS Code? If it's VS Code then would it be possible to construct a correct path using the VS Code server install location (which we should already have a variable for: `execPath`)? 

If it's the extension that updated then I see the issue. Maybe we could have an activation event for when an extension has been updated so that they can take care of things like this. I can think of other reasons an extension might want to activate on update too (migrations, deprecation messages, other warnings). This seems to match well with how the terminal substitution API works.

connor4312 commented:
> Was this they updated their extension, not VS Code? If it's VS Code then would it be possible to construct a correct path using the VS Code server install location (which we should already have a variable for: execPath)?

A VS Code update generally entails an extension update, and built-in extension now (it seems with the updating mechanism Sandeep added) are stored like ordinary extensions in a folder like `~/.vscode-server-insiders/<publisher>.<name>-<version>`. So we get a new path which each update.

> Maybe we could have an activation event for when an extension has been updated so that they can take care of things like this

I don't think this would work, the environment variable collections and workspace-specific and we'd need to activate and update them in each new workspace where they had previously been set.

alexr00 commented:
Thanks for answering all my questions! It's clear to me now that there isn't a better way to address this issue. 

Let's keep the new variable as publicly documented so that we don't have hidden API. When we do doc updates for 1.67 we need to make sure we add a sentence about how this variable can resolve to either a location on the remote or on the local machine depending on where the extension is installed.

alexr00 reviewed (changes requested):
Sorry, forgot one thing: to be consistent with the other variable the new one should probably be `extensionFolder` (https://code.visualstudio.com/docs/editor/variables-reference#_predefined-variables).

connor4312 commented:
Renamed! I added it to the variable completions where other variables were present.

Also, fixed a bug where the variable completions would only work for the first replacement per line.

alexr00 reviewed (approved):
Looks great, thank you!
